### PR TITLE
Semantic script workflow improved

### PR DESCRIPTION
**Pull request by denma98**
Regarding [issue ](https://github.com/tattle-made/feluda/issues/474#issue-2765500666),  I made the following improvements to the semantic_release_workflow.py script:

### Key Changes:

1. **Cached TOML Data**:
   -  Reduced redundant file reads by caching ```pyproject.toml``` data during package discovery. 
   - The _discover_packages method now parses ```pyproject.toml``` once and stores the data in ```self.packages```, which is reused across all functions.
3. **Early Validation**:
   -  Added checks for ```name```, ```version```, and ```tag_format``` in ```pyproject.toml``` during package discovery. 
   - Packages missing any of these fields are skipped immediately, preventing errors downstream.
5. **TOML Formatting**: 
   - Replaced ```tomli``` and ```tomli_w``` with ```tomlkit``` for parsing and writing ```pyproject.toml```.
   -  ```tomlkit``` preserves the original formatting, comments, and structure of the ```TOML``` file during updates, ensuring no unintended changes to the file layout.
7. **Breaking Change Detection**: 
   - Updated ```_parse_conventional_commit``` to scan the entire commit message for ```BREAKING CHANGE```. 
   - Ensures compliance with the Conventional Commits specification, where breaking changes can appear in the commit body.
9. **Git Command Fixes**: 
   - Improved handling of commit ranges in ```get_package_commits``` to avoid errors with root commits (no parent).
   -  Added error handling for Git commands to prevent crashes during edge cases.

### Testing:

- Added a comprehensive test script (```test_semantic_release.py```) to validate:

  - **No changes**: Ensures no version bump occurs when there are no changes.

  - **Patch bumps**: Verifies fix commits trigger a patch version bump (e.g., ```1.0.0``` → ```1.0.1```).

  - **Minor bumps**: Verifies feat commits trigger a minor version bump (e.g., ```1.0.0``` → ```1.1.0```).

  - **Major bumps**: Verifies breaking changes trigger a major version bump (e.g., ```1.0.0``` → ```2.0.0```).

  - **Tag handling**: Ensures version bumps are skipped if the tag for the new version already exists.
  

image

### CODE REVIEW

**Comment by denma98** [comment]
Hey @aatmanvaidya, please review this and check if some changes are needed.

**Comment by aatmanvaidya** [comment]
hi @denma98 this looks great, thank you!

please give me some time, I will review and get back to you!

**Comment by denma98** [comment]
sorry i had a folder name changed i have removed it now

**Comment by denma98** [comment]
> please give me some time, I will review and get back to you!

Sure, I will wait.

File: scripts/semantic_release_workflow.py
Comment (aatmanvaidya) [inline scripts/semantic_release_workflow.py:1]: hi @denma98 great work!
I really liked the optimisation's, neat work with adding `tomlkit`!
the test also looks great

I was testing out the changes made in the PR, and it is failing one case - so this script does semantic versioning for multiple packages in a monorepo. 

lets say there are 2 python packages in a repo called - `feluda` and `feluda-image-vec-rep-resnet`, now if I only made commits to the `feluda-image-vec-rep-resnet` folder, then the version of feluda-image-vec-rep-resnet should ONLY be changed and NOT feluda, 

right now what is happening is that the `feluda` version is also getting updated
here is the change I made to feluda-image-vec-rep-resnet - [commit](https://github.com/aatmanvaidya/feluda-versioning-check/commit/1466c66f3d99d36d71c87f1a828187d0e5ad353f) and here is what the script updated - [commit](https://github.com/aatmanvaidya/feluda-versioning-check/commit/7c486b451efba01f61b926394f26cd664c1d13fd), can you debug why this is happening? I did not take a very detailed look at the code

Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: Hey @aatmanvaidya , I changed the ```_discover_packages``` function and excluded the root directory to only look for subfolder packages, as you said. But should I ignore root completely or write code to change versions for feluda too whenever that happens?

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: yes yes you should ignore root completely, only if change is happening to feluda/ folder or its pyproject.toml file, then only its version should be updated

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: my old code logic cover's it well I think!

Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: I have updated the code to ignore root completely. Please check once

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: hi @denma98 I tested the new code, now this code is also not able to run semantic versioning when changes are made to `feluda`
this is change [I made from a PR](https://github.com/aatmanvaidya/feluda-versioning-check/commit/098e262362a3719c44e9bcc30ab28c74886244e7) - and here is the [github action results for it](https://github.com/aatmanvaidya/feluda-versioning-check/actions/runs/13922322208/job/38958239096), the code could not detect any changes to feluda and skipped semantic versioning, which should not happen. 

Also you can run some tests on your fork too, 
when you make the script changes, make sure you raise a PR with some changes and merge then, this I am guessing the action will also trigger on your end. 
I have written some script files to directly do everything also - https://github.com/aatmanvaidya/feluda-versioning-check/tree/main/automate_testing

Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: Okay, I'll look into it and try testing on my fork.

Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: Hey @aatmanvaidya, I made some changes to the code to update the ```get_package_commits``` method to exclude subdirectories when checking for changes to the root package (feluda). I made a PR request and merged it, and the [workflow](https://github.com/denma98/feluda-versioning-check/actions/runs/13971302363) for ```merge to PR``` ran well, showing success.
But the manual test for operators is failing, and I don't know why. Maybe I have to set up the ```GitHub.base_ref``` to detect changes.

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: hi @denma98 thank you, I will take a look soon, I am away for a bit of work, so please give me 2-3 days

what error is the manual test for operators is giving? what test are you running?

Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: This is the [operator ](https://github.com/denma98/feluda-versioning-check/actions/runs/13971332082) test I ran when merging the code to my main.  The workflow ran automatically when I merged the code to the main, but this test did not. So, I ran it manually. I am not sure why it is failing or do we even need to test it.

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: ahh okay, looks some error in the github action, so you can ignore it, all good!

Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: Okay

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: hi @denma98 so far these tests look good, I will re-take a detailed look again in the test and let you know

also, apologies for the delay in reviewing, I have been busy travelling, I will get back to you very soon!

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: hi @denma98 I noticed one more major bug
when I make a change JUST to an operator, semantic versioning is also getting applied to `feluda`, this should not be happening. 

Think of it like a multi-package python system in a monorepo. 

so here is the change I made - [link](https://github.com/aatmanvaidya/feluda-versioning-check/commit/cee40df35ad2fdd94df837538ee557d0d6a93309) - I have only made change to an opreator, but here was the output of the GitHub action - [link](https://github.com/aatmanvaidya/feluda-versioning-check/commit/d300262c6f905372bd8c9eb7f4f51b884144837d) - it made updates to the `feluda` root `pyproject.toml` file also

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: @denma98 I noticed you made a few commits, you can refer to my current code - [link](https://github.com/tattle-made/feluda/blob/main/scripts/semantic_release_workflow.py), it handles all such cases, 

I think you should just focus on minor optimisations and not changing the logic, this way you would also not have a hard time debugging where things are going wrong. The issue talks about fixing two main things
- make a fix to open pyproejct.toml file only once and lint/format it better (which you have figured out using `tomlikit`)
- add a function to check `name, version and tag_format` in a pyproject.toml file

so you can just make the above 2 changes in my code without any logic changes and you should be good to go I feel
do let me know if I can help out in any part!

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: @denma98 understood, 

yaa we should not be looking at the entire commit history, but a commit range that is a passed in the arguments of the script.
I will test things out on my end too

File: scripts/semantic_release_workflow.py [hunk 1]
```diff
@@ -4,8 +4,7 @@
 import subprocess
 import sys
 import tomlkit
-# import tomli
-# import tomli_w
+import semver
```
Comment (aatmanvaidya) [inline scripts/semantic_release_workflow.py:7]: @denma98 why is semver needed here? I thought I had written that logic from scratch (simply because I wanted that way as it gives us some more flexibility)

File: scripts/semantic_release_workflow.py [hunk 2]
```diff
@@ -160,40 +186,50 @@ def _bump_version(self, current_version, bump_type):
...
-
-            cmd = [
-                "git",
-                "log",
-                commit_range,
-                "--pretty=format:%B",  # Capture full commit message (subject + body)
-                "--",
-                package_path,
-            ]
+            if package_path == os.path.join(self.repo_root, "feluda"):
```
Comment (aatmanvaidya) [inline scripts/semantic_release_workflow.py:202]: is there a way to re-visit this if statement, because its making things a bit hard coded here, I would like to keep things hard coded only in the `discovery package` function as that's where we find all the folders and packages and not in `get package commits` functions.

File: scripts/semantic_release_workflow.py [hunk 3]
```diff
@@ -160,40 +186,50 @@ def _bump_version(self, current_version, bump_type):
...
+                cmd = [
+                    "git",
+                    "log",
+                    f"{self.prev_commit}..{self.current_commit}",
```
Comment (aatmanvaidya) [inline scripts/semantic_release_workflow.py:206]: also @denma98 can you kindly check if this line would work if we squash and merge a PR which has only one single commit, because what will happen here is that `self.prev_commit` and `self.current_commit` will be the same, this is when it fails I guess, that's why in my logic I had added this
https://github.com/tattle-made/feluda/blob/c6d5403567b00e219be796246534c66179976b61/scripts/semantic_release_workflow.py#L215-L222

notice the the `^` in the 
```python
f"{self.prev_commit}^..{self.current_commit}",
```

apologies for being very nit-picky with things, but I want to make sure we test for all scenarios, 

this workflow is a major part of feluda and breaking this will cause a lot of issues, hence want to be extra sure of things

File: scripts/semantic_release_workflow.py (hunk 1 above)
Comment (denma98) [reply scripts/semantic_release_workflow.py:7]: I was using it for debugging to ensure the version format was correct after changing the `g_parse_conventional_commit`. I forgot to remove that since it's already handled. I'll fix that.

File: scripts/semantic_release_workflow.py (hunk 2 above)
Comment (denma98) [reply scripts/semantic_release_workflow.py:202]: Maybe I can add a flag in the `discover_packages` function to indicate if a package requires special handling, such as including root files for Feluda. If the flag is set to true, I can also add `pyproject.toml` to the path check for the necessary components. Otherwise, if the package path is in operators, it will be handled by default.

File: scripts/semantic_release_workflow.py (hunk 3 above)
Comment (denma98) [reply scripts/semantic_release_workflow.py:206]: So, should I add this to the current code instead?  Or should I try merging the PR once?

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:206]: if you have the time, do test once on your end, else add this to current code instead

File: scripts/semantic_release_workflow.py (hunk 2 above)
Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:202]: i would let's keep it as it is right now, since there are a lot of moving part's, once we make the final fixes and merge this PR, we can take a call on this

File: scripts/semantic_release_workflow.py (hunk 3 above)
Comment (denma98) [reply scripts/semantic_release_workflow.py:210]: I have added this to the code. There was some error in merging the code 😞

File: scripts/semantic_release_workflow.py
Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: Also, do I need to resolve these conflicts on my end?

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:1]: no no, since we won't merge the PR here you won't have too

once I test it, you can directly open a PR in the main feluda repo, that's where we merge it

File: scripts/semantic_release_workflow.py [hunk 4]
```diff
@@ -160,40 +189,50 @@ def _bump_version(self, current_version, bump_type):
...
-                "--",
-                package_path,
-            ]
+            if package_path == os.path.join(self.repo_root, "feluda"):
+                cmd = [
+                    "git",
+                    "log",
+                    f"{self.prev_commit}^..{self.current_commit}",
+                    "--pretty=format:%s",
+                    "--full-history",
```
Comment (aatmanvaidya) [inline scripts/semantic_release_workflow.py:211]: @denma98 you will still have to remove the `--full-history` flag right? else the same error will happen?

File: scripts/semantic_release_workflow.py
Comment (denma98) [reply scripts/semantic_release_workflow.py:1]: Okay, understood. Thanks for the help.

**Comment by denma98** [comment]
> add the test to inside the `tests/` folder, create a new folder called `scripts/` and add the test there

Hey @aatmanvaidya, I have made changes to the Feluda fork and also changed the github workflow file. Regarding the test folder, should I add the ``test_semantic_script`` inside it?

**Comment by aatmanvaidya** [comment]
@denma98 yes inside the `tests` folder at root, create a folder called "scripts" inside it, put the test file
